ramips: add support for HILINK HLK-7628N

liuyuf78fk opened the pull request:
* ramips: add support for HILINK HLK-7628N

Specification:

CPU:       MT7628 580 MHz. MIPS 24K
RAM:       128 MB
Flash:     32 MB
WIFI:      802.11n/g/b 20/40 MHz
Ethernet:  5 Port ethernet switch
UART:      2x
LED:        1x wifi led
BUTTON: 1x reset button

Tested the following items：
Ethernet Switch
WIFI 
UART
WIFI LED
RESET BUTTON
SYSUPGRADE

Flash instruction:
The U-boot is based on Ralink SDK so we can flash the firmware using UART:
1. Configure PC with a static IP address and setup an TFTP server.
2. Put the firmware into the tftp directory.
3. Connect the UART0 line as described on the PCB.
4. Power up the device and press 2, follow the instruction to
   set device and tftp server IP address and input the firmware
   file name. U-boot will then load the firmware and write it into
   the flash.
5. After firmware is started connect via ethernet at 192.168.1.1

Signed-off-by: Liu Yu <f78fk@live.com>

liuyuf78fk commented:
root@OpenWrt:/# dmesg 
[    0.000000] Linux version 4.14.113 (liuyu@liuyu-virtual-machine) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r9902+1-9424b6f)) #0 Thu Apr 25 07:25:40 2019
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7628AN ver:1 eco:2
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019655 (MIPS 24KEc)
[    0.000000] MIPS: machine is HILINK HLK-7628N
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000] free_area_init_node: node 0, pgdat 80499240, node_mem_map 81000040
[    0.000000]   Normal zone: 256 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] random: get_random_bytes called from start_kernel+0x8c/0x474 with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32512
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=00023080
[    0.000000] Readback ErrCtl register=00023080
[    0.000000] Memory: 124696K/131072K available (3682K kernel code, 170K rwdata, 860K rodata, 208K init, 200K bss, 6376K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] intc: using register map from devicetree
[    0.000000] CPU Clock: 580MHz
[    0.000000] timer_probe: no matching timers found
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000013] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.007540] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.073496] pid_max: default: 32768 minimum: 301
[    0.078182] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.084530] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.097032] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.106529] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.112471] pinctrl core: initialized pinctrl subsystem
[    0.118577] NET: Registered protocol family 16
[    0.149185] mt7621_gpio 10000600.gpio: registering 32 gpios
[    0.154779] mt7621_gpio 10000600.gpio: registering 32 gpios
[    0.160394] mt7621_gpio 10000600.gpio: registering 32 gpios
[    0.170488] clocksource: Switched to clocksource MIPS
[    0.176530] NET: Registered protocol family 2
[    0.181588] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.188279] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.194467] TCP: Hash tables configured (established 1024 bind 1024)
[    0.200726] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.206328] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.212665] NET: Registered protocol family 1
[    0.216840] PCI: CLS 0 bytes, default 32
[    0.220190] Crashlog allocated RAM at address 0x3f00000
[    0.226832] workingset: timestamp_bits=30 max_order=15 bucket_order=0
[    0.238976] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.244601] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.263885] io scheduler noop registered
[    0.267617] io scheduler deadline registered (default)
[    0.273466] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    0.280835] console [ttyS0] disabled
[    0.284274] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 28, base_baud = 2500000) is a 16550A
[    0.293037] console [ttyS0] enabled
[    0.306675] bootconsole [early0] disabled
[    0.323148] 10000d00.uart1: ttyS1 at MMIO 0x10000d00 (irq = 29, base_baud = 2500000) is a 16550A
[    0.341226] 10000e00.uart2: ttyS2 at MMIO 0x10000e00 (irq = 30, base_baud = 2500000) is a 16550A
[    0.359196] cacheinfo: Failed to find cpu0 device node
[    0.369406] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.382231] spi-mt7621 10000b00.spi: sys_freq: 193333333
[    0.409584] m25p80 spi0.0: w25q256 (32768 Kbytes)
[    0.418982] 4 fixed-partitions partitions found on MTD device spi0.0
[    0.431572] Creating 4 MTD partitions on "spi0.0":
[    0.441073] 0x000000000000-0x000000030000 : "u-boot"
[    0.451867] 0x000000030000-0x000000040000 : "u-boot-env"
[    0.463342] 0x000000040000-0x000000050000 : "factory"
[    0.474285] 0x000000050000-0x000002000000 : "firmware"
[    0.488140] 2 uimage-fw partitions found on MTD device firmware
[    0.499928] Creating 2 MTD partitions on "firmware":
[    0.509772] 0x000000000000-0x000000182504 : "kernel"
[    0.520663] 0x000000182504-0x000001fb0000 : "rootfs"
[    0.531380] mtd: device 5 (rootfs) set to be root filesystem
[    0.544145] 1 squashfs-split partitions found on MTD device rootfs
[    0.556457] 0x000000380000-0x000001fb0000 : "rootfs_data"
[    0.568827] libphy: Fixed MDIO Bus: probed
[    0.587761] rt3050-esw 10110000.esw: link changed 0x00
[    0.598824] mtk_soc_eth 10100000.ethernet: generated random MAC address 0a:67:39:e8:08:3d
[    0.615812] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    0.634041] NET: Registered protocol family 10
[    0.646715] Segment Routing with IPv6
[    0.654162] NET: Registered protocol family 17
[    0.663030] 8021q: 802.1Q VLAN Support v1.8
[    0.680913] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    0.696034] Freeing unused kernel memory: 208K
[    0.704859] This architecture does not have kernel memory protection.
[    1.683711] init: Console is alive
[    1.690784] init: - watchdog -
[    2.275037] random: fast init done
[    2.740040] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.923343] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.948238] init: - preinit -
[    4.254737] rt3050-esw 10110000.esw: link changed 0x00
[    7.349102] rt3050-esw 10110000.esw: link changed 0x10
[    7.509842] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    7.549917] urandom-seed: Seed file not found (/etc/urandom.seed)
[    7.671359] procd: - early -
[    7.677166] procd: - watchdog -
[    8.323572] procd: - watchdog -
[    8.330094] procd: - ubus -
[    8.383424] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.397345] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.410677] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.424325] procd: - init -
[    8.938365] kmodloader: loading kernel modules from /etc/modules.d/*
[    8.958083] ip6_tables: (C) 2000-2006 Netfilter Core Team
[    8.978739] Loading modules backported from Linux version v4.19.32-0-g3a2156c839c7
[    8.993811] Backport generated by backports.git v4.19.32-1-0-g1c4f7569
[    9.010095] ip_tables: (C) 2000-2006 Netfilter Core Team
[    9.029682] nf_conntrack version 0.5.0 (2048 buckets, 8192 max)
[    9.103532] xt_time: kernel timezone is -0000
[    9.173307] mt76_wmac 10300000.wmac: ASIC revision: 76280001
[   10.202517] mt76_wmac 10300000.wmac: Firmware Version: 20151201
[   10.214330] mt76_wmac 10300000.wmac: Build Time: 20151201183641
[   10.240500] mt76_wmac 10300000.wmac: firmware init done
[   10.421020] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   10.534719] PPP generic driver version 2.4.2
[   10.546362] NET: Registered protocol family 24
[   10.559621] kmodloader: done loading kernel modules from /etc/modules.d/*
[   11.836425] urandom_read: 5 callbacks suppressed
[   11.836435] random: jshn: uninitialized urandom read (4 bytes read)
[   12.010734] random: jshn: uninitialized urandom read (4 bytes read)
[   23.134579] rt3050-esw 10110000.esw: link changed 0x00
[   25.211260] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   25.250739] jffs2_build_filesystem(): unlocking the mtd device... 
[   25.260703] done.
[   25.276766] jffs2_build_filesystem(): erasing all blocks after the end marker... 
[   26.329195] rt3050-esw 10110000.esw: link changed 0x10
[   29.106075] br-lan: port 1(eth0.1) entered blocking state
[   29.117031] br-lan: port 1(eth0.1) entered disabled state
[   29.128124] device eth0.1 entered promiscuous mode
[   29.137642] device eth0 entered promiscuous mode
[   29.274672] br-lan: port 1(eth0.1) entered blocking state
[   29.285429] br-lan: port 1(eth0.1) entered forwarding state
[   29.296687] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   30.090993] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   36.827524] rt3050-esw 10110000.esw: link changed 0x00
[  111.481089] random: crng init done

liuyuf78fk committed:
ramips: add support for HILINK HLK-7628N

* ramips: add support for HILINK HLK-7628N

Specification:

CPU:       MT7628 580 MHz. MIPS 24K
RAM:       128 MB
Flash:     32 MB
WIFI:      802.11n/g/b 20/40 MHz
Ethernet:  5 Port ethernet switch
UART:      2x

Flash instruction:
The U-boot is based on Ralink SDK so we can flash the firmware using UART:
1. Configure PC with a static IP address and setup an TFTP server.
2. Put the firmware into the tftp directory.
3. Connect the UART0 line as described on the PCB.
4. Power up the device and press 2, follow the instruction to
   set device and tftp server IP address and input the firmware
   file name. U-boot will then load the firmware and write it into
   the flash.
5. After firmware is started connect via ethernet at 192.168.1.1

Signed-off-by: Liu Yu <f78fk@live.com>

chunkeey commented on target/linux/ramips/dts/HLK-7628N.dts:
```diff
@@ -0,0 +1,92 @@
+/dts-v1/;
```
Please include licensing info for your DTS file, see
https://openwrt.org/submitting-patches#dts_checklist
e.g. a line like:
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
above the /dts-v1/;

chunkeey commented on target/linux/ramips/dts/HLK-7628N.dts:
```diff
@@ -0,0 +1,92 @@
+		bootargs = "console=ttyS0,57600";
+	};
+
+	gpio-keys-polled {
```
The name of a node should reflect the function of the device. Examples for generic node names can be found in Section 2.2.2 Generic Names Recommendation https://github.com/devicetree-org/devicetree-specification/blob/master/source/devicetree-basics.rst#generic-names-recommendation

chunkeey commented on target/linux/ramips/dts/HLK-7628N.dts:
```diff
@@ -0,0 +1,92 @@
+		};
+	};
+
+	gpio-leds {
```
Can you please rename the node to just "leds" ?

chunkeey commented on target/linux/ramips/dts/HLK-7628N.dts:
```diff
@@ -0,0 +1,92 @@
+&uart1 {
+	status = "okay";
+};
+&uart2 {
```
Is this uart2 connected to anything? If not, please remove the node.

chunkeey commented on target/linux/ramips/dts/HLK-7628N.dts:
```diff
@@ -0,0 +1,92 @@
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
```
Can you please rename this to "flash@0"?

liuyuf78fk replied:
> 你可以把它重命名为“flash @ 0”吗？

Already modified

981213 commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -188,6 +188,9 @@ hg255d)
 	ucidef_set_led_netdev "internet" "internet" "$boardname:green:internet" "eth0.2"
 	;;
+hilink,hlk-7628n)
+	set_wifi_led "hlk-7628n:green:wlan"
```
You should follow other examples using $boardname instead. This helps reducing script length as devices with similar LED naming scheme can be merged together.
It looks like you can already merge this one with these boards:
```bash
3g-6200nl|\
skylab,skw92a|\
wnce2001)
	set_wifi_led "$boardname:green:wlan"
	;;
```

liuyuf78fk replied:
> You should follow other examples using $boardname instead. This helps reducing script length as devices with similar LED naming scheme can be merged together.
> It looks like you can already merge this one with these boards:
> 
> ```shell
> 3g-6200nl|\
> skylab,skw92a|\
> wnce2001)
> 	set_wifi_led "$boardname:green:wlan"
> 	;;
> ```

Thanks for the suggestion，Already modified。

981213 reviewed (changes requested):
One more nitpicking :)

chunkeey commented:
@liuyuf78fk 
> Flash and recovery instructions: Use the factory installed u-boot boot
loader. It is available on UART0 (57600,8,n,1).  Then get the
sysupgrade image from a tftp server.

Do you think you could extend the u-boot instructions a bit? You can take stuff (copy and paste)
from [existing commits](https://git.openwrt.org/?p=openwrt/openwrt.git&a=search&h=HEAD&st=commit&s=add+support+for). For example, I would like to have the pinout of UART0 header and whenever it has TTL/CMOS or RS232-level (so people don't brick their SoCs). And maybe the exact commands you have to enter into the u-boot prompt (ideally in a way that they can be copy & pasted).

liuyuf78fk commented:
> @liuyuf78fk
> 
> > Flash and recovery instructions: Use the factory installed u-boot boot
> > loader. It is available on UART0 (57600,8,n,1).  Then get the
> > sysupgrade image from a tftp server.
> 
> Do you think you could extend the u-boot instructions a bit? You can take stuff (copy and paste)
> from [existing commits](https://git.openwrt.org/?p=openwrt/openwrt.git&a=search&h=HEAD&st=commit&s=add+support+for). For example, I would like to have the pinout of UART0 header and whenever it has TTL/CMOS or RS232-level (so people don't brick their SoCs). And maybe the exact commands you have to enter into the u-boot prompt (ideally in a way that they can be copy & pasted).



> @liuyuf78fk
> 
> > Flash and recovery instructions: Use the factory installed u-boot boot
> > loader. It is available on UART0 (57600,8,n,1).  Then get the
> > sysupgrade image from a tftp server.
> 
> Do you think you could extend the u-boot instructions a bit? You can take stuff (copy and paste)
> from [existing commits](https://git.openwrt.org/?p=openwrt/openwrt.git&a=search&h=HEAD&st=commit&s=add+support+for). For example, I would like to have the pinout of UART0 header and whenever it has TTL/CMOS or RS232-level (so people don't brick their SoCs). And maybe the exact commands you have to enter into the u-boot prompt (ideally in a way that they can be copy & pasted).

Already modified。

chunkeey commented:
Merged. Thanks.

adrianschmutzler commented:
@liuyuf78fk 
Hi, I'm currently reviewing the MAC address assignment on ramips target.
This device is not setting up a MAC address for &ethernet at all in DTS (your dmesg also reports that address is random).
Can you provide a hexdump of your factory partition, so we can work out the correct MAC address setup for that device?

alex9x commented:
Hi! Does someone know, why don`t work software reset on HLK-7628N? 
4.xx have patch for 32mb flash with 0x66 0x99 sequence to back flash in 3byte address mode.
18.06.9 works well, but master or 19.x.x based on 5.x kernel fail to make soft reset.
Any help are welcome!

liuyuf78fk commented:
> @liuyuf78fk
> Hi, I'm currently reviewing the MAC address assignment on ramips target.
> This device is not setting up a MAC address for &ethernet at all in DTS (your dmesg also reports that address is random).
> Can you provide a hexdump of your factory partition, so we can work out the correct MAC address setup for that device?

Hello, I have modified it and added a pull request.

liuyuf78fk commented:
> Hi! Does someone know, why don`t work software reset on HLK-7628N?
> 4.xx have patch for 32mb flash with 0x66 0x99 sequence to back flash in 3byte address mode.
> 18.06.9 works well, but master or 19.x.x based on 5.x kernel fail to make soft reset.
> Any help are welcome!
I found that this is a bug in SPI nor flash driver
https://github.com/openwrt/openwrt/pull/4248
